README
How to compile GFAL 2.0 ?


- general compilation :
cd [gfal_dir]
scons 

- compile and execute tests :
cd [gfal_dir]
scons test

- compile and execute examples  :
cd [gfal_dir]
scons testing


- generates all RPMS  :
cd [gfal_dir]
./package_generator.sh





RELEASE-NOTES
*** Release glite-data-gfal_R_1_11_16_2 2010-11-15
    Fixed bugs:
        #75340: lcg-get-checksum crashes when srmLs does not return checksum

    Fixed bugs:
        #74869: lcg-get-checksum crashes in case of invalid proxy
        #74459: lcg-gt - segfault

*** Release glite-data-gfal_R_1_11_15_5 2010-09-13
    Fixed bugs:
        #72275: Problems when executing CTB tests
        #70844: [lcg_util] lcg-del fails when --nobdii argument is passed

    o Certification tests imported and added to the automated test execution

*** Release glite-data-gfal_R_1_11_15_1 2010-04-15
    Fixed bugs: 
        #64518: Make sources EGEE copyright and license compliant
        #62444: [GFAL-Python] gfal_get and gfal_prestage return (-1) on success
        #52502: gfal_mkdir and DPM SE
        #61445: [lcg_util] Wrong file size registered in LFC when reading named pipe

*** Release glite-data-gfal_R_1_11_14_1 2010-01-25

    Fixed bugs:
        #61765: [lcg_util] lcg_del segfault when there is no user mapping
        #60991: lcg-cr gives vague error for existing LFN
        
*** Release glite-data-gfal_R_1_11_13_1 2009-12-18

    Fixed bugs:
    	#60334: [lcg_util] lcg-cp hangs when file has no replica
	    #55424: gfal file open problem(gfal_open)

*** Release glite-data-gfal_R_1_11_12_2 2009-11-24

	o fixed the build of the unit test

*** Release glite-data-gfal_R_1_11_12_1 2009-11-20

    Fixed bugs:
        #53141: [GFAL] gfal_abortrequest() TypeError: iteration over non-sequence 
        #54172: Errno is not set when gfal_opendir_python fails  
        #54849: GFAL: shall handle abreviated checksum names 
        #55266: GFAL man page for gfal_get_results and gfal_get_results_python needs updating 
        #55576: RFE: Extra parameter in lcg-cp for a better TURL construction 
        #56373: [GFAL] memory corruption problem with classic SEs 
        #56950: GFAL's src/Makefile.am is missing CGSI_GSOAP_CFLAGS 
	o Fix some memory leaks
    o Improved testability and added unit tests
    o Code refactorzation/reindentation.
	o Change LICENSE to Apache2

*** Release glite-data-gfal_R_1_11_11_1 2009-09-14
    o Fix bug #55266: GFAL man page for gfal_get_results and 
      gfal_get_results_python needs updating
    o Fix bug #41844: IPv6 bug; LCG-utils client 
      functionality immediately broken by IPv6
    
*** Release glite-data-gfal_R_1_11_10_1 2009-08-26
    o Fix bug #54849: GFAL: shall handle abreviated checksum names

*** Release glite-data-gfal_R_1_11_9_1 2009-08-14
    o Fix bug #54259: Unable to get a tURL in a full space
    o Fix bug #54333: GFAL: problems with LDAP queries on SL5
    o Error message in case of wrongly published or missing purt info in BDII

*** Release glite-data-gfal_R_1_11_8_1 2009-07-13

    o Fix bug #53122: [GFAL] gfal_pin() - result file status = 1 (instead of 0) after successful request
    o Fix compilation warnings in SWIG-generated code
    o Fix bug #52893: [GFAL] gfal_ls errno value is wrong
    o Fix bug #52872: [GFAL] gfal_stat errno value is wrong
    o Fix bug #52838: [GFAL] gfal_lseek TypeError: Expected a pointer
    o Fix some compilation (harmless) warnings
    o Fix bug #46066: GFAL: allowing info discovery in site BDII (as well)
    o Fix bug #45455: GFAL BDII failover can be improved
    o Fix bug #52714: [GFAL] wrong errno value return code for gfal_chmod
    o Fix bug #52545: [GFAL] wrong errno value and return code for gfal_unlink
    o Fix bug #52645: code of error EFAULT is wrong
    o Fix bug #52476: Swig error message with Python 2.5 on SL5
    o Fix bug #52485: lcg-cp fails to copy to gsiftp TURL

*** Release glite-data-gfal_R_1_11_7_2 2009-06-24

    o We can use dcache-dcap as build dependency on all platforms.

*** Release glite-data-gfal_R_1_11_7_1 2009-06-05

    o Fix bug #51230: [GFAL] gfal_open segfaults on <connection refused>
    o Fix bug #51179: [GFAL] lcg-cr issues one lfc_starttrans but two lfc_endtrans calls

*** Release glite-data-gfal_R_1_11_6_2 2009-05-20
    o Fixing some ETICS locking problems.

*** Release glite-data-gfal_R_1_11_6_1 2009-05-19

    o Fix bug #50613: [GFAL] error handling regression with gfal_ls

*** Release glite-data-gfal_R_1_11_5_1 2009-04-22

    o Fix bug #47606: [gfal] checksum-related field have to be returned in python API
    o Fix bug #47605: File size is not set after gfal_close.
    o Fix bug #47493: [lcg-utils] a warning should be printed if command line VO doesn't match VOMS proxy one
    o Fix bug #48185: support for async srmLs
    o Fix bug #47122: [GFAL] add a new function for bringonline with polling/retrying mechanism
    o Fix bug #48940: [GFAL] Pinlifetime not returned for BringOnline requests
    o Fix bug #48815: [lcg-utils] lcg-del fails deleting LFN with no replica
    o Fix bug #48571: [GFAL] Freeing static pointer returned by get_default_se() in gfal_open()
    o Fix bug #47026: [lcg_util] RFE: lcg-get-checksum
    o Fix bug #47508: Please, make SRM v2 the default in GFAL/lcg-utils
    o Fix bug #47120: [GFAL] add srm return code in error message

*** Release glite-data-gfal_R_1_11_4_1 2009-02-17

    o Fix bug #46561: [gfal] "file exists" error on already existing directories when creating a new file

*** Release glite-data-gfal_R_1_11_2_1 2009-01-30

    o Fix bug #46409: [GFAL] recursive directory creation should be changed
    o Fix bug #44637: lcg-cr with GFAL-client-1.10.17 Fail to create directories in SToRM SE

*** Release glite-data-gfal_R_1_11_1_1 2009-01-26

    o Add release number with autotools
    o Fix bug #46276: [GFAL] 'extern "C" {' isn't closed in gfal_types.h and gfal_internals.h

*** Release glite-data-gfal_R_1_11_0_1 2008-12-19

    o Change python from lib/python to /lib/python2.X/site-packages
    o Introduce new function gfal_get_errno to get errno value with python API
    o Add python function gfal_listdir (equivalent of os.listdir python function)
    o Prints a message saying that RMC/LRC are obsolete

src/se_ifce.c
/*
 * Copyright (C) 2005-2006 by CERN
 */

/*
 * @(#)$RCSfile: se_ifce.c,v $ $Revision: 1.13 $ $Date: 2007/06/07 13:56:14 $ CERN Jean-Philippe Baud
 */

#include <sys/types.h>
#include <errno.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <sys/stat.h>
#include "gfal_api.h"
#undef SOAP_FMAC3
#define SOAP_FMAC3 static
#undef SOAP_FMAC5
#define SOAP_FMAC5 static
#include "seH.h"
#include "edg_se_webserviceSoapBinding+.nsmap"
#ifdef GFAL_SECURE
#include "cgsi_plugin.h"
#endif
#include "seC.c"
#include "seClient.c"

static int errorstring2errno (const char *);

se_deletesurls (int nbfiles, const char **surls, const char *srm_endpoint,
		struct se_filestatus **filestatuses, char *errbuf, int errbufsz, int timeout)
{
	int ret;
	int flags;
	struct ns1__deleteResponse out;
	struct soap soap;
	int i;

	if ((*filestatuses = (struct se_filestatus *) calloc (nbfiles, sizeof (struct se_filestatus))) == NULL) {
		errno = ENOMEM;
		return (-1);
	}

	soap_init (&soap);
	soap.namespaces = namespaces_se;
#ifdef GFAL_SECURE
	flags = CGSI_OPT_DISABLE_NAME_CHECK | CGSI_OPT_SSL_COMPATIBLE;
	soap_register_plugin_arg (&soap, client_cgsi_plugin, &flags);
#endif

	soap.send_timeout = timeout ;
	soap.recv_timeout = timeout ;

	/* issue "delete" requests */

	for (i = 0; i < nbfiles; ++i) {
		bzero (*filestatuses + i, sizeof (struct se_filestatus));
		(*filestatuses)[i].surl = strdup (surls[i]);

		if ((ret = soap_call_ns1__delete (&soap, srm_endpoint,
					"delete", (char *)(surls[i] + 6), &out))) {
			char errmsg[ERRMSG_LEN];
			if (soap.error == SOAP_EOF) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*filestatuses)[i].status = ECOMM;
			} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*filestatuses)[i].status = errorstring2errno (soap.fault->faultstring);
			} else {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*filestatuses)[i].status = ECOMM;
			}
		}
	}
	soap_end (&soap);
	soap_done (&soap);
	return (nbfiles);
}

se_mkdir (const char *dir, const char *srm_endpoint, char *errbuf, int errbufsz, int timeout)
{
	int flags;
	struct ns1__mkdirResponse out;
	int ret;
	int sav_errno;
	struct soap soap;

	soap_init (&soap);
	soap.namespaces = namespaces_se;
#ifdef GFAL_SECURE
	flags = CGSI_OPT_DISABLE_NAME_CHECK | CGSI_OPT_SSL_COMPATIBLE;
	soap_register_plugin_arg (&soap, client_cgsi_plugin, &flags);
#endif

	soap.send_timeout = timeout ;
	soap.recv_timeout = timeout ;

	/* issue "mkdir" request */
	if ((ret = soap_call_ns1__mkdir (&soap, srm_endpoint,
					"mkdir", (char *)dir + 6, &out))) {
		char errmsg[ERRMSG_LEN];
		if (soap.error == SOAP_EOF) {
			snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
			gfal_errmsg(errbuf, errbufsz, errmsg);
			sav_errno = ECOMM;
		} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
			snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
			gfal_errmsg(errbuf, errbufsz, errmsg);
			sav_errno = errorstring2errno (soap.fault->faultstring);
		} else {
			snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
			gfal_errmsg(errbuf, errbufsz, errmsg);
			sav_errno = ECOMM;
		}
		soap_end (&soap);
		soap_done (&soap);
		errno = sav_errno;
		return (-1);
	}
	soap_end (&soap);
	soap_done (&soap);
	return (0);
}

static int
se_makedirp (const char *surl, const char *srm_endpoint, char *errbuf, int errbufsz, int timeout)
{
	int c;
	char *lastslash = NULL;
	char *p;
	char *p1;
	char *q;
	char sav_surl[1104];
	struct se_mdfilestatus *mdstatuses;

	if (strlen (surl) >= sizeof(sav_surl)) {
		char errmsg[ERRMSG_LEN];
		snprintf (errmsg, ERRMSG_LEN - 1, "%s: SURL too long", surl);
		gfal_errmsg(errbuf, errbufsz, errmsg);
		errno = ENAMETOOLONG;
		return (-1);
	}
	strcpy (sav_surl, surl);
	p1 = strchr (sav_surl + 6, '/');
	p = strrchr (sav_surl + 6, '/');
	while (p > p1) {
		if (lastslash == NULL) lastslash = p;
		*p = '\0';
		c = se_getfilemd (1, (const char **) &sav_surl, srm_endpoint, &mdstatuses, errbuf, errbufsz, timeout);
		if (mdstatuses[0].surl) free (mdstatuses[0].surl);
		if (mdstatuses[0].status != ENOENT) {
			errno = mdstatuses[0].status;
			free (mdstatuses);
			return (c);
		}
		free (mdstatuses);
		if (c == 0) {
			*p = '/';
			break;
		}
		q = strrchr (sav_surl, '/');
		*p = '/';
		p = q;
	}
	c = 0;
	while (c == 0 && (p = strchr (p + 1, '/')) && p <= lastslash) {
		*p = '\0';
		c = se_mkdir (sav_surl, srm_endpoint, errbuf, errbufsz, timeout);
		*p = '/';
	}
	return (c);
}

se_turlsfromsurls (int nbfiles, const char **surls, const char *srm_endpoint, char **protocols, int oflag,
		struct se_filestatus **filestatuses, char *errbuf, int errbufsz, int timeout)
{
	int flags;
	int nbproto = 0;
	struct ns1__cacheResponse outg;
	struct ns1__createResponse outp;
	struct ns1__getTurlResponse outq;
	char *p;
	struct ArrayOf_USCORExsd_USCOREstring protoarray;
	int ret, i;
	int sav_errno;
	struct soap soap;
	LONG64 zero = 0;

	if ((*filestatuses = (struct se_filestatus *) calloc (nbfiles, sizeof (struct se_filestatus))) == NULL) {
		errno = ENOMEM;
		return (-1);
	}

	soap_init (&soap);
	soap.namespaces = namespaces_se;
#ifdef GFAL_SECURE
	flags = CGSI_OPT_DISABLE_NAME_CHECK | CGSI_OPT_SSL_COMPATIBLE;
	soap_register_plugin_arg (&soap, client_cgsi_plugin, &flags);
#endif

	soap.send_timeout = timeout ;
	soap.recv_timeout = timeout ;

	while (*protocols[nbproto]) nbproto++;

	/* issue "cache" or the "create" request */

	protoarray.__ptr = protocols;
	protoarray.__size = nbproto;

	for (i = 0; i < nbfiles; ++i) {
		bzero (*filestatuses + i, sizeof (struct se_filestatus));
		(*filestatuses)[i].surl = strdup (surls[i]);

		if ((oflag & O_ACCMODE) == 0) {
			if ((ret = soap_call_ns1__cache (&soap, srm_endpoint, "cache",
							(char *)(surls[i] + 6), "read", 36000, &outg))) {
				char errmsg[ERRMSG_LEN];
				if (soap.error == SOAP_EOF) {
					snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
					gfal_errmsg(errbuf, errbufsz, errmsg);
					(*filestatuses)[i].status = ECOMM;
				} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
					snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
					gfal_errmsg(errbuf, errbufsz, errmsg);
					(*filestatuses)[i].status = errorstring2errno (soap.fault->faultstring);
				} else {
					snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
					gfal_errmsg(errbuf, errbufsz, errmsg);
					(*filestatuses)[i].status = ECOMM;
				}
				continue;
			}

			if (((*filestatuses)[i].token = strdup (outg._cacheReturn)) == NULL) {
				(*filestatuses)[i].status = ENOMEM;
				continue;
			}
		} else {
			if (se_makedirp (surls[i], srm_endpoint, errbuf, errbufsz, timeout) == 0) {
				(*filestatuses)[i].status = errno;
				errno = 0;
				continue;
			}
			if ((ret = soap_call_ns1__create (&soap, srm_endpoint, "create",
							(char *)(surls[i] + 6), zero, "x", 36000, &outp))) {
				char errmsg[ERRMSG_LEN];
				if (soap.error == SOAP_EOF) {
					snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
					gfal_errmsg(errbuf, errbufsz, errmsg);
					(*filestatuses)[i].status = ECOMM;
				} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
					snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
					gfal_errmsg(errbuf, errbufsz, errmsg);
					(*filestatuses)[i].status = errorstring2errno (soap.fault->faultstring);
				} else {
					snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
					gfal_errmsg(errbuf, errbufsz, errmsg);
					(*filestatuses)[i].status = ECOMM;
				}
				continue;
			}
			if (((*filestatuses)[i].token = strdup (outp._createReturn)) == NULL) {
				(*filestatuses)[i].status = ENOMEM;
				continue;
			}
		}

		if ((ret = soap_call_ns1__getTurl (&soap, srm_endpoint, "getTurl", (*filestatuses)[i].token,
						&protoarray, &outq))) {
			char errmsg[ERRMSG_LEN];
			if (soap.error == SOAP_EOF) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*filestatuses)[i].status = ECOMM;
			} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*filestatuses)[i].status = errorstring2errno (soap.fault->faultstring);
			} else {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*filestatuses)[i].status = ECOMM;
			}
			continue;
		}

		(*filestatuses)[i].turl = strdup (outq._getTurlReturn);
	}

	soap_end (&soap);
	soap_done (&soap);
	return (nbfiles);
}

se_getfilemd (int nbfiles, const char **surls, const char *srm_endpoint,
		struct se_mdfilestatus **mdstatuses, char *errbuf, int errbufsz, int timeout)
{
	int flags;
	char *dp;
	struct group *gr;
	int i, j;
	struct ns1__getMetadataResponse out;
	char *p;
	struct passwd *pw;
	char *q;
	int ret;
	struct soap soap;

	if ((*mdstatuses = (struct se_mdfilestatus *) calloc (nbfiles, sizeof (struct se_mdfilestatus))) == NULL) {
		errno = ENOMEM;
		return (-1);
	}

	soap_init (&soap);
	soap.namespaces = namespaces_se;
#ifdef GFAL_SECURE
	flags = CGSI_OPT_DISABLE_NAME_CHECK | CGSI_OPT_SSL_COMPATIBLE;
	soap_register_plugin_arg (&soap, client_cgsi_plugin, &flags);
#endif

	soap.send_timeout = timeout;
	soap.recv_timeout = timeout;

	/* issue "getMetadata" request */

	for (i = 0; i < nbfiles; ++i) {
		bzero (*mdstatuses + i, sizeof (struct se_mdfilestatus));
		(*mdstatuses)[i].surl = strdup (surls[i]);

		if ((ret = soap_call_ns1__getMetadata (&soap, srm_endpoint,
						"getMetadata", (char *)(surls[i] + 6), &out))) {
			char errmsg[ERRMSG_LEN];
			if (soap.error == SOAP_EOF) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*mdstatuses)[i].status = ECOMM;
			} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*mdstatuses)[i].status = errorstring2errno (soap.fault->faultstring);
			} else {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				(*mdstatuses)[i].status = ECOMM;
			}
			continue;
		}

		(*mdstatuses)[i].stat.st_nlink = 1;
		(*mdstatuses)[i].stat.st_uid = 2;
		(*mdstatuses)[i].stat.st_gid = 2;
		(*mdstatuses)[i].stat.st_mode = 0666;
		for (j = 0; j < out._getMetadataReturn->__size; j++) {
			p = out._getMetadataReturn->__ptr[j];
			if ((q = strchr (p, ':'))) {
				*q = '\0';
				if (strcmp (p, "filesize") == 0)
					(*mdstatuses)[i].stat.st_size = strtoll (q + 1, &dp, 10);
				else if (strcmp (p, "atime") == 0)
					(*mdstatuses)[i].stat.st_atime = strtol (q + 1, &dp, 10);
				else if (strcmp (p, "mtime") == 0)
					(*mdstatuses)[i].stat.st_mtime = strtol (q + 1, &dp, 10);
				else if (strcmp (p, "ctime") == 0)
					(*mdstatuses)[i].stat.st_ctime = strtol (q + 1, &dp, 10);
			}
		}
	}
	soap_end (&soap);
	soap_done (&soap);
	return (0);
}

se_set_xfer_done (const char *srm_endpoint, const char *token,
		int oflag, char *errbuf, int errbufsz, int timeout)
{
	int flags;
	int ret;
	struct soap soap;
	int sav_errno;

	soap_init (&soap);
	soap.namespaces = namespaces_se;
#ifdef GFAL_SECURE
	flags = CGSI_OPT_DISABLE_NAME_CHECK | CGSI_OPT_SSL_COMPATIBLE;
	soap_register_plugin_arg (&soap, client_cgsi_plugin, &flags);
#endif

	soap.send_timeout = timeout ;
	soap.recv_timeout = timeout ;


	if ((oflag & O_ACCMODE) == 0) {
		/*	not implemented yet 
		struct ns1__abandonResponse outa;

		if (soap_call_ns1__abandon (&soap, srm_endpoint,
					"abandon", token, &outa)) {
			gfal_errmsg(errbuf, errbufsz, soap.fault->faultstring);
			soap_end (&soap);
			soap_done (&soap);
			return (-1);
		}
		*/
	} else {
		struct ns1__commitResponse outc;

		if ((ret = soap_call_ns1__commit (&soap, srm_endpoint,
					"commit", (char *) token, &outc))) {
			char errmsg[ERRMSG_LEN];
			if (soap.error == SOAP_EOF) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Connection fails or timeout", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				sav_errno = ECOMM;
			} else if(ret == SOAP_FAULT || ret == SOAP_CLI_FAULT) {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: %s", srm_endpoint, soap.fault->faultstring);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				sav_errno = errorstring2errno (soap.fault->faultstring);
			} else {
				snprintf (errmsg, ERRMSG_LEN - 1, "%s: Unknown error", srm_endpoint);
				gfal_errmsg(errbuf, errbufsz, errmsg);
				sav_errno = ECOMM;
			}
			soap_end (&soap);
			soap_done (&soap);
			errno = sav_errno;
			return (-1);
		}
	}
	soap_end (&soap);
	soap_done (&soap);
	return (0);
}

se_set_xfer_running (const char *srm_endpoint, const char *token,
		char *errbuf, int errbufsz, int timeout)
{
	return (0);
}

static int
errorstring2errno (const char *errstr)
{
	if (!errstr)
		return (ECOMM);
	else if (strstr (errstr, "ile exists"))
		return (EEXIST);
	else if (strstr (errstr, "does not exist") ||
			strstr (errstr, "o such file") ||
			strstr (errstr, "not found") ||
			strstr (errstr, "could not get storage info by path"))
		return (ENOENT);
	else if (strstr (errstr, "ermission denied") ||
			strstr (errstr, "expired"))
		return (EACCES);
	else if (strstr (errstr, "nvalid arg"))
		return (EINVAL);
	else if (strstr (errstr, "rotocol"))
		return (EPROTONOSUPPORT);
	else if (strstr (errstr, "o space left on device"))
		return (ENOSPC);
	else
		return (ECOMM);
}
